src/renderers/shared/fiber/ReactFiberContext.js

```diff
@@ -117,21 +122,34 @@ function processChildContext(fiber : Fiber, parentContext : Object, isReconcilin
 }
 exports.processChildContext = processChildContext;
 
-exports.pushContextProvider = function(workInProgress : Fiber, didPerformWork : boolean) : void {
-  const instance = workInProgress.stateNode;
-  const memoizedMergedChildContext = instance.__reactInternalMemoizedMergedChildContext;
-  const canReuseMergedChildContext = !didPerformWork && memoizedMergedChildContext != null;
-
-  let mergedContext = null;
-  if (canReuseMergedChildContext) {
-    mergedContext = memoizedMergedChildContext;
-  } else {
-    mergedContext = processChildContext(workInProgress, getUnmaskedContext(), true);
-    instance.__reactInternalMemoizedMergedChildContext = mergedContext;
+exports.pushContextProvider = function(workInProgress : Fiber) : boolean {
+  if (!isContextProvider(workInProgress)) {
+    return false;
   }
 
-  push(contextStackCursor, mergedContext, workInProgress);
-  push(didPerformWorkStackCursor, didPerformWork, workInProgress);
+  const instance = workInProgress.stateNode;
+  // We push the context as early as possible to ensure stack integrity.
+  // If the instance does not exist yet, we will push null at first,
+  // and replace it on the stack later when invalidating the context.
+  const memoizedMergedChildContext = (
+    instance &&
+    instance.__reactInternalMemoizedMergedChildContext
+  ) || null;
+  push(contextStackCursor, memoizedMergedChildContext, workInProgress);
+  push(didPerformWorkStackCursor, false, workInProgress);
+  return true;
+};
+
+exports.invalidateContextProvider = function(workInProgress : Fiber) : void {
+  const instance = workInProgress.stateNode;
+  if (instance == null) {
+    throw new Error('Expected to have an instance by this point.');
```

Comment:
Curious why you're using `throw` here instead of `invariant`? (How do we determine when to do one vs the other?)